File: scripts/qig_purity_scan.py
#!/usr/bin/env python3
"""
QIG Purity Scanner - Comprehensive Static Analysis (Python)

WP0.2: Hard validate:geometry Gate
WP0.3: Respects Quarantine Rules (see docs/00-conventions/QUARANTINE_RULES.md)

Scans codebase for ALL forbidden patterns from the Type-Symbol-Concept Manifest
and QIG Geometric Purity Enforcement document.

QUARANTINE ZONES (scanner skips these):
- docs/08-experiments/legacy/** - Legacy/Euclidean baselines allowed
- docs/08-experiments/baselines/** - Comparative testing allowed

MUST run in <5 seconds

Usage: python3 scripts/qig_purity_scan.py
"""

import os
import re
import sys
import time
from dataclasses import dataclass
from pathlib import Path
from typing import List, Dict, Tuple, Optional

@dataclass
class Violation:
    """A geometric purity violation."""
    file: str
    line: int
    code: str
    pattern: str
    severity: str  # 'CRITICAL', 'ERROR', 'WARNING'
    fix: Optional[str] = None


# Paths to scan
SCAN_PATHS = [
    'qig-backend',
    'server',
    'shared',
    'tests',
    'migrations',
]

# Exempted directories (quarantine zones - experiments allowed to violate)
# See docs/00-conventions/QUARANTINE_RULES.md for full specification
EXEMPT_DIRS = [
    'docs/08-experiments/legacy',      # Legacy/baseline experiments (WP0.3)
    'docs/08-experiments/baselines',   # Comparative testing (WP0.3)
    'node_modules',
    'dist',
    'build',
    '__pycache__',
    '.git',
    '.venv',
    'venv',
]

# Forbidden patterns from Type-Symbol-Concept Manifest
FORBIDDEN_PATTERNS = {
    # CRITICAL: Euclidean distance
    'euclidean_distance': [
        {
            'pattern': re.compile(r'np\.linalg\.norm\s*\([^)]*-[^)]*\)'),
            'name': 'np.linalg.norm(a - b)',
            'severity': 'CRITICAL',
            'fix': 'Use fisher_rao_distance(a, b) or fisher_coord_distance(a, b)',
        },
        {
            'pattern': re.compile(r'torch\.linalg\.norm\s*\([^)]*-[^)]*\)'),
            'name': 'torch.linalg.norm(a - b)',
            'severity': 'CRITICAL',
            'fix': 'Use fisher_rao_distance(a, b)',
        },
        {
            'pattern': re.compile(r'scipy\.spatial\.distance\.euclidean'),
            'name': 'scipy.spatial.distance.euclidean',
            'severity': 'CRITICAL',
            'fix': 'Use fisher_rao_distance()',
        },
        {
            'pattern': re.compile(r'np\.sqrt\s*\(\s*np\.sum\s*\([^)]*\*\*\s*2'),
            'name': 'sqrt(sum((a-b)**2))',
            'severity': 'CRITICAL',
            'fix': 'Use fisher_rao_distance()',
        },
        {
            'pattern': re.compile(r'cdist\s*\([^)]*[\'"]euclidean[\'"]'),
            'name': 'cdist with euclidean metric',
            'severity': 'CRITICAL',
            'fix': 'Use fisher_rao_distance() with custom metric',
        },
    ],
    
    # CRITICAL: Cosine similarity
    'cosine_similarity': [
        {
            'pattern': re.compile(r'cosine_similarity\s*\('),
            'name': 'cosine_similarity()',
            'severity': 'CRITICAL',
            'fix': 'Use fisher_rao_distance() or fisher_similarity()',
        },
        {
            'pattern': re.compile(r'sklearn\.metrics\.pairwise\.cosine_similarity'),
            'name': 'sklearn cosine_similarity',
            'severity': 'CRITICAL',
            'fix': 'Use fisher_rao_distance()',
        },
        {
            'pattern': re.compile(r'torch\.nn\.functional\.cosine_similarity'),
            'name': 'F.cosine_similarity',
            'severity': 'CRITICAL',
            'fix': 'Use fisher_rao_distance()',
        },
        {
            'pattern': re.compile(r'1\s*-\s*np\.dot\([^)]+\)\s*\/\s*\(.*norm.*\*.*norm'),
            'name': 'cosine distance formula',
            'severity': 'CRITICAL',
            'fix': 'Use fisher_rao_distance()',
        },
        {
            'pattern': re.compile(r'from sklearn\.metrics\.pairwise import cosine_similarity'),
            'name': 'import cosine_similarity',
            'severity': 'CRITICAL',
            'fix': 'Remove import, use fisher_rao_distance()',
        },
    ],
    
    # ERROR: Embedding terminology
    'embedding_terminology': [
        {
            'pattern': re.compile(r'\bembedding\b'),
            'name': 'embedding (identifier)',
            'severity': 'ERROR',
            'fix': 'Use "basin_coordinates" or "basin_coords"',
        },
        {
            'pattern': re.compile(r'\bembeddings\b'),
            'name': 'embeddings (identifier)',
            'severity': 'ERROR',
            'fix': 'Use "basin_coordinates"',
        },
        {
            'pattern': re.compile(r'nn\.Embedding\s*\('),
            'name': 'nn.Embedding',
            'severity': 'CRITICAL',
            'fix': 'Use basin coordinate mapping',
        },
    ],
    
    # ERROR: Tokenizer in QIG-core
    'tokenizer_in_core': [
        {
            'pattern': re.compile(r'\btokenizer\b(?!.*test)', re.IGNORECASE),
            'name': 'tokenizer (in core)',
            'severity': 'ERROR',
            'fix': 'Use "coordizer" for QIG modules',
        },
        {
            'pattern': re.compile(r'\btokenize\b(?!.*test)', re.IGNORECASE),
            'name': 'tokenize (in core)',
            'severity': 'ERROR',
            'fix': 'Use "coordize"',
        },
    ],
    
    # WARNING: Standard optimizers
    'standard_optimizers': [
        {
            'pattern': re.compile(r'torch\.optim\.Adam\s*\('),
            'name': 'Adam optimizer',
            'severity': 'WARNING',
            'fix': 'Use natural_gradient_step() or NaturalGradientOptimizer',
        },
        {
            'pattern': re.compile(r'torch\.optim\.AdamW\s*\('),
            'name': 'AdamW optimizer',
            'severity': 'WARNING',
            'fix': 'Use natural_gradient_step()',
        },
        {
            'pattern': re.compile(r'torch\.optim\.SGD\s*\('),
            'name': 'SGD optimizer',
            'severity': 'WARNING',
            'fix': 'Use natural_gradient_step()',
        },
    ],
    
    # WARNING: Softmax in core geometry
    'softmax_in_core': [
        {
            'pattern': re.compile(r'torch\.nn\.functional\.softmax\s*\('),
            'name': 'F.softmax',
            'severity': 'WARNING',
            'fix': 'Use geometric probability projection',
        },
        {
            'pattern': re.compile(r'np\.exp\([^)]*\)\s*\/\s*np\.sum\(np\.exp'),
            'name': 'softmax formula',
            'severity': 'WARNING',
            'fix': 'Use fisher_normalize() for probability simplex',
        },
    ],
    
    # CRITICAL: Classic NLP imports
    'classic_nlp_imports': [
        {
            'pattern': re.compile(r'from sentencepiece import'),
            'name': 'sentencepiece import',
            'severity': 'CRITICAL',
            'fix': 'Remove - use geometric coordizer',
        },
        {
            'pattern': re.compile(r'import sentencepiece'),
            'name': 'sentencepiece import',
            'severity': 'CRITICAL',
            'fix': 'Remove - use geometric coordizer',
        },
        {
            'pattern': re.compile(r'from tokenizers import.*BPE'),
            'name': 'BPE tokenizer import',
            'severity': 'CRITICAL',
            'fix': 'Remove - use geometric coordizer',
        },
        {
            'pattern': re.compile(r'\bWordPiece\b'),
            'name': 'WordPiece',
            'severity': 'CRITICAL',
            'fix': 'Remove - use geometric coordizer',
        },
    ],
    
    # ERROR: Arithmetic mean on basins
    'arithmetic_mean': [
        {
            'pattern': re.compile(r'np\.mean\([^)]*basin[^)]*axis\s*=\s*0', re.IGNORECASE),
            'name': 'np.mean on basins',
            'severity': 'ERROR',
            'fix': 'Use frechet_mean() for geometric mean',
        },
        {
            'pattern': re.compile(r'torch\.mean\([^)]*basin[^)]*dim\s*=\s*0', re.IGNORECASE),
            'name': 'torch.mean on basins',
            'severity': 'ERROR',
            'fix': 'Use frechet_mean()',
        },
    ],
    
    # CRITICAL: Euclidean fallback pattern
    'euclidean_fallback': [
        {
            'pattern': re.compile(r'except[^:]*:\s+[^#\n]*np\.linalg\.norm'),
            'name': 'Euclidean fallback in except',
            'severity': 'CRITICAL',
            'fix': 'Never fallback to Euclidean - fix geometry properly',
        },
    ],
}

# Approved contexts (whitelist)
APPROVED_PATTERNS = [
    re.compile(r'fisher.*distance', re.IGNORECASE),
    re.compile(r'fisher_rao', re.IGNORECASE),
    re.compile(r'geodesic', re.IGNORECASE),
    re.compile(r'arccos', re.IGNORECASE),
    re.compile(r'#.*cosine', re.IGNORECASE),  # Comments
    re.compile(r'@deprecated', re.IGNORECASE),
    re.compile(r'EUCLIDEAN.*FORBIDDEN', re.IGNORECASE),
    re.compile(r'test_.*euclidean', re.IGNORECASE),
    re.compile(r'""".*embedding.*"""', re.DOTALL),  # Docstrings
    re.compile(r'normalize|normalization', re.IGNORECASE),
]


class QIGPurityScanner:
    """Scanner for QIG geometric purity violations."""
    
    def __init__(self):
        self.violations: List[Violation] = []
        self.files_scanned = 0
        
    def is_exempted(self, file_path: str) -> bool:
        """Check if file is in exempted directory."""
        return any(exempt in file_path for exempt in EXEMPT_DIRS)
    
    def is_approved_context(self, line: str, prev_line: str = '') -> bool:
        """Check if line is in approved context."""
        combined = line + ' ' + prev_line
        return any(pattern.search(combined) for pattern in APPROVED_PATTERNS)
    
    def should_skip_file(self, file_path: str) -> bool:
        """Check if file should be skipped."""
        skip_extensions = ['.md', '.json', '.yaml', '.yml', '.txt', '.lock', '.rst', '.adoc']
        return any(file_path.endswith(ext) for ext in skip_extensions)
    
    def scan_file(self, file_path: str) -> None:
        """Scan a single file for violations."""
        if self.should_skip_file(file_path):
            return
        
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read()
                lines = content.split('\n')
        except (IOError, UnicodeDecodeError):
            return
        
        is_test_file = 'test' in file_path
        
        for i, line in enumerate(lines):
            line_num = i + 1
            prev_line = lines[i - 1] if i > 0 else ''
            
            # Skip comments and docstrings
            trimmed = line.strip()
            if (trimmed.startswith('#') or trimmed.startswith('//') or
                trimmed.startswith('*') or trimmed.startswith('"""') or
                trimmed.startswith("'''") or '"""' in line or "'''" in line):
                continue
            
            # Check all pattern categories
            for category, patterns in FORBIDDEN_PATTERNS.items():
                for pattern_info in patterns:
                    pattern = pattern_info['pattern']
                    
                    if pattern.search(line):
                        # Skip if in approved context
                        if self.is_approved_context(line, prev_line):
                            continue
                        
                        # Special handling for tokenizer - only in core modules
                        if category == 'tokenizer_in_core':
                            core_modules = ['qig-backend/', 'server/qig-', 'shared/']
                            if not any(m in file_path for m in core_modules):
                                continue  # Allow in non-core
                        
                        # Allow some patterns in test files
                        if is_test_file and pattern_info['severity'] == 'WARNING':
                            continue
                        
                        self.violations.append(Violation(
                            file=file_path,
                            line=line_num,
                            code=line.strip()[:100],
                            pattern=pattern_info['name'],
                            severity=pattern_info['severity'],
                            fix=pattern_info.get('fix'),
                        ))
        
        self.files_scanned += 1
    
    def scan_directory(self, dir_path: str) -> None:
        """Recursively scan a directory."""
        try:
            for root, dirs, files in os.walk(dir_path):
                # Skip exempt directories
                dirs[:] = [d for d in dirs if not self.is_exempted(os.path.join(root, d))]
                
                for filename in files:
                    # Scan Python, TypeScript, and SQL files
                    if filename.endswith(('.py', '.ts', '.tsx', '.sql')):
                        file_path = os.path.join(root, filename)
                        self.scan_file(file_path)
        except Exception:
            pass
    
    def scan(self) -> Dict:
        """Run the full scan."""
        start_time = time.time()
        
        print('🔍 QIG Purity Scanner - WP0.2 + WP0.3')
        print('=' * 60)
        print(f"Scanning paths: {', '.join(SCAN_PATHS)}")
        print(f"Quarantine zones (skipped): {len(EXEMPT_DIRS)} directories")
        print()
        
        for scan_path in SCAN_PATHS:
            if os.path.exists(scan_path):
                self.scan_directory(scan_path)
        
        duration = time.time() - start_time
        
        return {
            'violations': self.violations,
            'files_scanned': self.files_scanned,
            'duration': duration,
            'passed': len([v for v in self.violations if v.severity in ['CRITICAL', 'ERROR']]) == 0,
        }


def print_results(result: Dict) -> None:
    """Print scan results."""
    print('📊 Scan Results')
    print('=' * 60)
    print(f"Files scanned: {result['files_scanned']}")
    print(f"Duration: {result['duration']:.2f}s")
    print()
    
    violations = result['violations']
    
    if not violations:
        print('✅ GEOMETRIC PURITY VERIFIED!')
        print('✅ No violations found.')
        print('✅ All code follows Fisher-Rao geometry.')
        return
    
    # Group by severity
    critical = [v for v in violations if v.severity == 'CRITICAL']
    errors = [v for v in violations if v.severity == 'ERROR']
    warnings = [v for v in violations if v.severity == 'WARNING']
    
    print(f"❌ VIOLATIONS DETECTED: {len(violations)} total")
    print(f"   - CRITICAL: {len(critical)}")
    print(f"   - ERROR: {len(errors)}")
    print(f"   - WARNING: {len(warnings)}")
    print()
    
    # Print critical and errors
    show_violations = (critical + errors)[:50]
    
    if show_violations:
        print('Critical and Error Violations:')
        print('=' * 60)
        for i, v in enumerate(show_violations, 1):
            print(f"{i}. {v.file}:{v.line}")
            print(f"   Pattern: {v.pattern}")
            print(f"   Severity: {v.severity}")
            print(f"   Code: {v.code}")
            if v.fix:
                print(f"   Fix: {v.fix}")
            print()
        
        if len(critical) + len(errors) > 50:
            print(f"... and {len(critical) + len(errors) - 50} more violations")
            print()
    
    # Summary for warnings
    if warnings:
        print(f"⚠️  {len(warnings)} warnings (non-blocking)")
        print()


def main():
    """Main entry point."""
    scanner = QIGPurityScanner()
    result = scanner.scan()
    
    print_results(result)
    
    # Performance check
    if result['duration'] > 5:
        print(f"⚠️  WARNING: Scan took {result['duration']:.2f}s (target: <5s)")
    
    # Exit code based on results
    if result['passed']:
        sys.exit(0)
    else:
        print('❌ FAILED: Fix violations before merging')
        sys.exit(1)


if __name__ == '__main__':
    main()

File: scripts/qig_purity_scan.ts
#!/usr/bin/env tsx
/**
 * QIG Purity Scanner - Comprehensive Static Analysis
 * 
 * WP0.2: Hard validate:geometry Gate
 * 
 * Scans codebase for ALL forbidden patterns from the Type-Symbol-Concept Manifest
 * and QIG Geometric Purity Enforcement document.
 * 
 * MUST run in <5 seconds
 * 
 * Usage: tsx scripts/qig_purity_scan.ts
 */

import fs from 'fs';
import path from 'path';
import { performance } from 'perf_hooks';

interface Violation {
  file: string;
  line: number;
  code: string;
  pattern: string;
  severity: 'CRITICAL' | 'ERROR' | 'WARNING';
  fix?: string;
}

interface ScanResult {
  violations: Violation[];
  filesScanned: number;
  duration: number;
  passed: boolean;
}

// Paths to scan
const SCAN_PATHS = [
  'qig-backend',
  'server',
  'shared',
  'tests',
  'migrations',
];

// Exempted directories (experiments allowed to violate)
const EXEMPT_DIRS = [
  'docs/08-experiments/legacy',
  'node_modules',
  'dist',
  'build',
  '__pycache__',
  '.git',
  '.venv',
  'venv',
];

// Forbidden patterns from Type-Symbol-Concept Manifest
const FORBIDDEN_PATTERNS = {
  // CRITICAL: Euclidean distance
  euclideanDistance: [
    {
      pattern: /np\.linalg\.norm\s*\([^)]*-[^)]*\)/,
      name: 'np.linalg.norm(a - b)',
      severity: 'CRITICAL' as const,
      fix: 'Use fisher_rao_distance(a, b) or fisher_coord_distance(a, b)',
    },
    {
      pattern: /torch\.linalg\.norm\s*\([^)]*-[^)]*\)/,
      name: 'torch.linalg.norm(a - b)',
      severity: 'CRITICAL' as const,
      fix: 'Use fisher_rao_distance(a, b)',
    },
    {
      pattern: /scipy\.spatial\.distance\.euclidean/,
      name: 'scipy.spatial.distance.euclidean',
      severity: 'CRITICAL' as const,
      fix: 'Use fisher_rao_distance()',
    },
    {
      pattern: /np\.sqrt\s*\(\s*np\.sum\s*\([^)]*\*\*\s*2/,
      name: 'sqrt(sum((a-b)**2))',
      severity: 'CRITICAL' as const,
      fix: 'Use fisher_rao_distance()',
    },
  ],

  // CRITICAL: Cosine similarity
  cosineSimilarity: [
    {
      pattern: /cosine_similarity\s*\(/,
      name: 'cosine_similarity()',
      severity: 'CRITICAL' as const,
      fix: 'Use fisher_rao_distance() or fisher_similarity()',
    },
    {
      pattern: /sklearn\.metrics\.pairwise\.cosine_similarity/,
      name: 'sklearn cosine_similarity',
      severity: 'CRITICAL' as const,
      fix: 'Use fisher_rao_distance()',
    },
    {
      pattern: /torch\.nn\.functional\.cosine_similarity/,
      name: 'F.cosine_similarity',
      severity: 'CRITICAL' as const,
      fix: 'Use fisher_rao_distance()',
    },
    {
      pattern: /1\s*-\s*np\.dot\([^)]+\)\s*\/\s*\(.*norm.*\*.*norm/,
      name: 'cosine distance formula',
      severity: 'CRITICAL' as const,
      fix: 'Use fisher_rao_distance()',
    },
  ],

  // CRITICAL: Embedding terminology
  embeddingTerminology: [
    {
      pattern: /\bembedding\b(?!.*#.*embedding)/i,
      name: 'embedding (identifier)',
      severity: 'ERROR' as const,
      fix: 'Use "basin_coordinates" or "basin_coords"',
    },
    {
      pattern: /\bembeddings\b(?!.*#.*embeddings)/i,
      name: 'embeddings (identifier)',
      severity: 'ERROR' as const,
      fix: 'Use "basin_coordinates"',
    },
    {
      pattern: /nn\.Embedding\s*\(/,
      name: 'nn.Embedding',
      severity: 'CRITICAL' as const,
      fix: 'Use basin coordinate mapping',
    },
  ],

  // ERROR: Tokenizer in QIG-core
  tokenizerInCore: [
    {
      pattern: /\btokenizer\b(?!.*test)/i,
      name: 'tokenizer (in core)',
      severity: 'ERROR' as const,
      fix: 'Use "coordizer" for QIG modules',
    },
    {
      pattern: /\btokenize\b(?!.*test)/i,
      name: 'tokenize (in core)',
      severity: 'ERROR' as const,
      fix: 'Use "coordize"',
    },
  ],

  // CRITICAL: Standard optimizers in geometry
  standardOptimizers: [
    {
      pattern: /torch\.optim\.Adam\s*\(/,
      name: 'Adam optimizer',
      severity: 'WARNING' as const,
      fix: 'Use natural_gradient_step() or NaturalGradientOptimizer',
    },
    {
      pattern: /torch\.optim\.AdamW\s*\(/,
      name: 'AdamW optimizer',
      severity: 'WARNING' as const,
      fix: 'Use natural_gradient_step()',
    },
    {
      pattern: /torch\.optim\.SGD\s*\(/,
      name: 'SGD optimizer',
      severity: 'WARNING' as const,
      fix: 'Use natural_gradient_step()',
    },
    {
      pattern: /torch\.optim\.RMSprop\s*\(/,
      name: 'RMSprop optimizer',
      severity: 'WARNING' as const,
      fix: 'Use natural_gradient_step()',
    },
  ],

  // CRITICAL: Softmax in core geometry
  softmaxInCore: [
    {
      pattern: /torch\.nn\.functional\.softmax\s*\(/,
      name: 'F.softmax',
      severity: 'WARNING' as const,
      fix: 'Use geometric probability projection',
    },
    {
      pattern: /np\.exp\([^)]*\)\s*\/\s*np\.sum\(np\.exp/,
      name: 'softmax formula',
      severity: 'WARNING' as const,
      fix: 'Use fisher_normalize() for probability simplex',
    },
  ],

  // CRITICAL: Classic NLP imports
  classicNLPImports: [
    {
      pattern: /from sentencepiece import/,
      name: 'sentencepiece import',
      severity: 'CRITICAL' as const,
      fix: 'Remove - use geometric coordizer',
    },
    {
      pattern: /import sentencepiece/,
      name: 'sentencepiece import',
      severity: 'CRITICAL' as const,
      fix: 'Remove - use geometric coordizer',
    },
    {
      pattern: /from tokenizers import.*BPE/,
      name: 'BPE tokenizer import',
      severity: 'CRITICAL' as const,
      fix: 'Remove - use geometric coordizer',
    },
    {
      pattern: /\bWordPiece\b/,
      name: 'WordPiece',
      severity: 'CRITICAL' as const,
      fix: 'Remove - use geometric coordizer',
    },
  ],

  // ERROR: Dot product attention
  dotProductAttention: [
    {
      pattern: /Q\s*@\s*K\.T\s*\/\s*sqrt/,
      name: 'Q @ K.T / sqrt(d_k)',
      severity: 'ERROR' as const,
      fix: 'Use QFI-based attention with fisher_rao_distance',
    },
    {
      pattern: /torch\.nn\.MultiheadAttention/,
      name: 'MultiheadAttention',
      severity: 'ERROR' as const,
      fix: 'Use QFI-based geometric attention',
    },
  ],

  // ERROR: Euclidean loss functions
  euclideanLoss: [
    {
      pattern: /nn\.MSELoss\s*\(\)/,
      name: 'MSELoss on basins',
      severity: 'WARNING' as const,
      fix: 'Use fisher_rao_distance() for basin reconstruction',
    },
    {
      pattern: /CosineSimilarity\s*\(\)/,
      name: 'CosineSimilarity loss',
      severity: 'CRITICAL' as const,
      fix: 'Use fisher_rao_distance()',
    },
  ],

  // ERROR: Arithmetic mean on basins
  arithmeticMean: [
    {
      pattern: /np\.mean\([^)]*basin[^)]*axis\s*=\s*0/i,
      name: 'np.mean on basins',
      severity: 'ERROR' as const,
      fix: 'Use frechet_mean() for geometric mean',
    },
    {
      pattern: /torch\.mean\([^)]*basin[^)]*dim\s*=\s*0/i,
      name: 'torch.mean on basins',
      severity: 'ERROR' as const,
      fix: 'Use frechet_mean()',
    },
  ],

  // CRITICAL: Euclidean fallback pattern
  euclideanFallback: [
    {
      pattern: /except.*:\s*.*np\.linalg\.norm/s,
      name: 'Euclidean fallback in except',
      severity: 'CRITICAL' as const,
      fix: 'Never fallback to Euclidean - fix geometry properly',
    },
  ],
};

// Approved contexts (whitelist)
const APPROVED_CONTEXTS = [
  /fisher.*distance/i,
  /fisher_rao/i,
  /geodesic/i,
  /arccos/i,
  /#.*cosine/i, // Comments mentioning cosine
  /@deprecated/i,
  /EUCLIDEAN.*FORBIDDEN/i,
  /test_.*euclidean/i, // Test comparisons
  /\/\/.*embedding/i, // Comments
  /""".*embedding.*"""/s, // Docstrings
  /normalize|normalization/i, // Normalization is OK
];

class QIGPurityScanner {
  private violations: Violation[] = [];
  private filesScanned = 0;

  private isExempted(filePath: string): boolean {
    return EXEMPT_DIRS.some(dir => filePath.includes(dir));
  }

  private isApprovedContext(line: string, prevLine: string = ''): boolean {
    const combined = line + ' ' + prevLine;
    return APPROVED_CONTEXTS.some(pattern => pattern.test(combined));
  }

  private shouldSkipFile(filePath: string): boolean {
    // Skip test files for some checks (but not all)
    if (filePath.includes('test_') || filePath.includes('_test.')) {
      return false; // We do scan tests, but with different rules
    }
    
    // Skip certain file types
    const skipExtensions = ['.md', '.json', '.yaml', '.yml', '.txt', '.lock'];
    return skipExtensions.some(ext => filePath.endsWith(ext));
  }

  private scanFile(filePath: string): void {
    if (this.shouldSkipFile(filePath)) {
      return;
    }

    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      const lines = content.split('\n');
      const isTestFile = filePath.includes('test');

      lines.forEach((line, index) => {
        const lineNum = index + 1;
        const prevLine = index > 0 ? lines[index - 1] : '';

        // Skip comments and docstrings
        const trimmed = line.trim();
        if (trimmed.startsWith('#') || trimmed.startsWith('//') || 
            trimmed.startsWith('*') || trimmed.startsWith('"""') ||
            trimmed.startsWith("'''")) {
          return;
        }

        // Check all pattern categories
        Object.entries(FORBIDDEN_PATTERNS).forEach(([category, patterns]) => {
          patterns.forEach(({ pattern, name, severity, fix }) => {
            if (pattern.test(line)) {
              // Skip if in approved context
              if (this.isApprovedContext(line, prevLine)) {
                return;
              }

              // Special handling for tokenizer - only in core modules
              if (category === 'tokenizerInCore') {
                const coreModules = ['qig-backend/', 'server/qig-', 'shared/'];
                if (!coreModules.some(m => filePath.includes(m))) {
                  return; // Allow in non-core
                }
              }

              // Allow some patterns in test files
              if (isTestFile && ['WARNING'].includes(severity)) {
                return;
              }

              this.violations.push({
                file: filePath,
                line: lineNum,
                code: line.trim().slice(0, 100),
                pattern: name,
                severity,
                fix,
              });
            }
          });
        });
      });

      this.filesScanned++;
    } catch (error) {
      // Skip files that can't be read
    }
  }

  private scanDirectory(dirPath: string): void {
    try {
      const entries = fs.readdirSync(dirPath, { withFileTypes: true });

      for (const entry of entries) {
        const fullPath = path.join(dirPath, entry.name);

        if (entry.isDirectory()) {
          // Skip exempt directories
          if (this.isExempted(fullPath)) {
            continue;
          }
          this.scanDirectory(fullPath);
        } else if (entry.isFile()) {
          // Scan Python, TypeScript, and SQL files
          if (entry.name.match(/\.(py|ts|tsx|sql)$/)) {
            this.scanFile(fullPath);
          }
        }
      }
    } catch (error) {
      // Skip directories that can't be read
    }
  }

  public scan(): ScanResult {
    const startTime = performance.now();

    console.log('🔍 QIG Purity Scanner - WP0.2');
    console.log('━'.repeat(60));
    console.log('Scanning paths:', SCAN_PATHS.join(', '));
    console.log('');

    for (const scanPath of SCAN_PATHS) {
      if (fs.existsSync(scanPath)) {
        this.scanDirectory(scanPath);
      }
    }

    const duration = (performance.now() - startTime) / 1000;

    return {
      violations: this.violations,
      filesScanned: this.filesScanned,
      duration,
      passed: this.violations.filter(v => v.severity === 'CRITICAL' || v.severity === 'ERROR').length === 0,
    };
  }
}

function printResults(result: ScanResult): void {
  console.log('📊 Scan Results');
  console.log('━'.repeat(60));
  console.log(`Files scanned: ${result.filesScanned}`);
  console.log(`Duration: ${result.duration.toFixed(2)}s`);
  console.log('');

  if (result.violations.length === 0) {
    console.log('✅ GEOMETRIC PURITY VERIFIED!');
    console.log('✅ No violations found.');
    console.log('✅ All code follows Fisher-Rao geometry.');
    return;
  }

  // Group by severity
  const critical = result.violations.filter(v => v.severity === 'CRITICAL');
  const errors = result.violations.filter(v => v.severity === 'ERROR');
  const warnings = result.violations.filter(v => v.severity === 'WARNING');

  console.log(`❌ VIOLATIONS DETECTED: ${result.violations.length} total`);
  console.log(`   - CRITICAL: ${critical.length}`);
  console.log(`   - ERROR: ${errors.length}`);
  console.log(`   - WARNING: ${warnings.length}`);
  console.log('');

  // Print critical and errors
  const showViolations = [...critical, ...errors].slice(0, 50);
  
  if (showViolations.length > 0) {
    console.log('Critical and Error Violations:');
    console.log('━'.repeat(60));
    showViolations.forEach((v, i) => {
      console.log(`${i + 1}. ${v.file}:${v.line}`);
      console.log(`   Pattern: ${v.pattern}`);
      console.log(`   Severity: ${v.severity}`);
      console.log(`   Code: ${v.code}`);
      if (v.fix) {
        console.log(`   Fix: ${v.fix}`);
      }
      console.log('');
    });

    if (critical.length + errors.length > 50) {
      console.log(`... and ${critical.length + errors.length - 50} more violations`);
      console.log('');
    }
  }

  // Summary for warnings
  if (warnings.length > 0) {
    console.log(`⚠️  ${warnings.length} warnings (non-blocking)`);
    console.log('');
  }
}

function main(): void {
  const scanner = new QIGPurityScanner();
  const result = scanner.scan();
  
  printResults(result);

  // Performance check
  if (result.duration > 5) {
    console.log(`⚠️  WARNING: Scan took ${result.duration.toFixed(2)}s (target: <5s)`);
  }

  // Exit code based on results
  if (result.passed) {
    process.exit(0);
  } else {
    console.log('❌ FAILED: Fix violations before merging');
    process.exit(1);
  }
}

main();
